pkg/agent/memberlist/server.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/klog"
+)
+
+type Server struct {
+	serverVersion   string
+	bindPort        int
+	nodeConfig      *config.NodeConfig
+	nodeInformer    coreinformers.NodeInformer
+	memberList      *memberlist.Memberlist
+	existingMembers *[]string
+}
+
+func NewMemberlistServer(p int, nodeInformer coreinformers.NodeInformer, nodeConfig *config.NodeConfig) *Server {
+	klog.Infof("Node config: %#v", nodeConfig)
+
+	s := &Server{
+		bindPort:      p,
+		serverVersion: "v1",
+		nodeInformer:  nodeInformer,
+		nodeConfig:    nodeConfig,
+	}
+
+	hostname := s.nodeConfig.Name
+	bindPort := s.bindPort
+	hostIP := s.nodeConfig.NodeIPAddr.IP
+
+	nodeMember := fmt.Sprintf("%s:%d", hostIP.String(), bindPort)
+
+	klog.Infof("Add new node: %s", nodeMember)
+
+	conf := memberlist.DefaultLocalConfig()
+	conf.Name = hostname + "-" + strconv.Itoa(bindPort)
+
+	conf.BindPort = bindPort
+	conf.AdvertisePort = bindPort
+
+	klog.Infof("Configs: %+v\n", conf)
+
+	list, err := memberlist.Create(conf)
+	if err != nil {
+		panic("Failed to create memberlist: " + err.Error())
+	}
+
+	s.memberList = list
+	s.existingMembers = &[]string{nodeMember}
+
+	// Join an existing cluster by specifying at least one known member.
+	s.JoinMembers(*s.existingMembers)
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    s.addStatleAgentCRD,
+		UpdateFunc: nil,
+		DeleteFunc: nil,
+	})
+
+	return s
+}
+
+func (ms *Server) convertListNodesToMemberlist() []string {
+	nodes, err := ms.nodeInformer.Lister().List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error when listing Nodes: %v", err)
+	}
+	klog.Infof("List %d nodes.", len(nodes))
+
+	clusterNodes := make([]string, len(nodes))
+
+	for i, node := range nodes {
+		klog.Infof("node %s: %#v", node.Name, node.Status.Addresses)
+		address := node.Status.Addresses
+		for _, add := range address {
+			if add.Type == corev1.NodeInternalIP {
+				member := fmt.Sprintf("%s:%d", add.Address, ms.bindPort)
+				clusterNodes[i] = member
+				klog.Infof("Cluster memberlist: %s.", member)
+			}
+		}
+	}
+	return clusterNodes
+}
+
+func (ms *Server) addStatleAgentCRD(old interface{}) {
+	node, ok := old.(*corev1.Node)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Node, invalid type: %v", old)
+			return
+		}
+		node, ok = tombstone.Obj.(*corev1.Node)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Node, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	ms.addMember(node)
+}
+
+func (ms *Server) addMember(node *corev1.Node) {
+	var member string
+	for _, add := range node.Status.Addresses {
+		if add.Type == corev1.NodeInternalIP {
+			member = fmt.Sprintf("%s:%d", add.Address, ms.bindPort)
+		}
+	}
+	if member != "" {
+		*ms.existingMembers = append(*ms.existingMembers, member)
+		ms.JoinMembers(*ms.existingMembers)
+	}
+}
+
+func (ms *Server) JoinMembers(clusterNodes []string) {
```

Comment:
make methods public only when they are required

Reply:
hi, @tnqn , Thanks so much for review. I will check these code details as you mentioned.
well, I want to ask for sugestion that whether the general direction is right. Then I can focus on details and unit tests.